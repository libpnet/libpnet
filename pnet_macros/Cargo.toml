[package]

name = "pnet_macros"
version = "0.0.1"
authors = ["Robert Clipsham <robert@octarineparrot.com>"]
license = "MIT/Apache-2.0"
homepage = "https://github.com/libpnet/libpnet"
repository = "https://github.com/libpnet/libpnet"
documentation = "http://octarineparrot.com/assets/libpnet/doc/pnet/"
description = "Macros for libpnet"
readme = "../README.md"
keywords = ["networking", "transport", "datalink", "packet", "protocol"]

[lib]
name = "pnet_macros"
plugin = true

[dependencies.pnet]
path = ".."
version = "*"
