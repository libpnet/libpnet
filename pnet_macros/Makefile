# Path to compiletest binary (compile from rustc/src/compiletest)
COMPILETEST=compiletest
# Path to rustc
RUSTC=rustc
# pnet_macros library path
PNET_MACROS=$(shell ls target/libpnet_macros*.so)
# pnet library path
PNET=$(shell ls target/deps/libpnet*.rlib)
# Target triple
TRIPLE=x86_64-unknown-linux-gnu
# Args for compiletest
COMPILETEST_ARGS=--compile-lib-path "" --run-lib-path "" --rustc-path $(RUSTC) --src-base tests \
				 --build-base target/tests --aux-base tests/aux --stage-id "" \
				 --android-cross-path "" --target $(TRIPLE) --target-rustcflags \
				 "--extern pnet=$(PNET) --extern pnet_macros=$(PNET_MACROS)"

test:
	$(COMPILETEST) --mode compile-fail $(COMPILETEST_ARGS) compile-fail
