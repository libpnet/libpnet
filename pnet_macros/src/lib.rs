// Copyright (c) 2015 Robert Clipsham <robert@octarineparrot.com>
// Copyright (c) 2021 Pierre Chifflier <chifflier@wzdftpd.net>
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

//! The pnet_macros crate provides the `#[packet]` macro and compiler plugin, which is used to
//! specify the format of on-the-wire packets, and automatically generate zero-copy accessors and
//! mutators for the fields. It is used as follows:
//!
//! ```
//! /// Import the #[packet] custom attribute
//! use pnet_macros::packet;
//! /// This module contains a list of type aliases which may be used
//! use pnet_macros_support::types::{u4, u12be};
//!
//! /// Packets are specified in the same way as normal Rust structs, but with a `#[packet]`
//! /// attribute.
//! #[packet]
//! pub struct Example {
//!     // This is a simple field which contains a 4-bit, unsigned integer.
//!     // Note that `u4` is simply an alias for `u8` - the name is a hint
//!     // to the compiler plugin, it is NOT a usable 4 bit type!
//!     simple_field1: u4,
//!
//!     // This specifies that `simple_field2` should be a 12-bit field,
//!     // with bits stored in big endian
//!     simple_field2: u12be,
//!
//!     // All packets must specify a `#[payload]`, which should be a
//!     // `Vec<u8>`. This represents the packet's payload, for example in
//!     // an IPv4 packet, the payload could be a UDP packet, or in a UDP
//!     // packet the payload could be the application data. All the
//!     // remaining space in the packet is considered to be the payload
//!     // (this doesn't have to be the case, see the documentation for
//!     // `#[payload]` below.
//!     #[payload]
//!     payload: Vec<u8>
//! }
//! ```
//! A number of things will then be generated. You can see this in action in the documentation and
//! source of each of the packet types in the `pnet::packet` module. Things generated include
//! (assuming the `Example` struct from above):
//!
//!  * An `ExamplePacket<'p>` structure, which is used for receiving packets on the network.
//!    This structure contains:
//!      - A method, `pub fn new<'p>(packet: &'p [u8]) -> ExamplePacket<'p>`, used for the
//!        construction of an `ExamplePacket`, given a buffer to store it. The buffer should be
//!        long enough to contain all the fields in the packet.
//!      - A method, `pub fn to_immutable<'p>(&'p self) -> ExamplePacket<'p>`, which is simply an
//!        identity function. It exists for consistency with `MutableExamplePacket`.
//!      - A number of accessor methods, of the form `pub get_{field_name}(&self) -> {field_type}`,
//!        which will retrieve the host representation of the on-the-wire value.
//!  * A `MutableExamplePacket<'p>` structure, which is used when sending packets on the network.
//!    This structure contains:
//!      - A method, `pub fn new<'p>(packet: &'p mut [u8]) -> MutableExamplePacket<'p>`, used for
//!        the construction of a `MutableExamplePacket`, given a buffer to store it. The buffer
//!        should be long enough to contain all the fields in the packet.
//!      - A method, `pub fn to_immutable<'p>(&'p self) -> ExamplePacket<'p>`, which converts from
//!        a `MutableExamplePacket` to an `ExamplePacket`
//!      - A method, `pub fn populate(&mut self, packet: Example)`, which, given an `Example`
//!        struct, will populate the `MutableExamplePacket` with the values from the `Example`
//!        struct.
//!      - A number of accessor methods, of the form `pub get_{field_name}(&self) -> {field_type}`,
//!        which will retrieve the host representation of the on-the-wire value.
//!      - A number of mutator methods, of the form `pub set_{field_name}(&mut self,
//!        val: {field_type})`, which will take a host value, convert it to the required
//!        on-the-wire format, and store it in the buffer which backs the `MutableExamplePacket`.
//!  * A number of trait implementations for each of the `MutableExamplePacket` and `ExamplePacket`
//!    structures. These include:
//!      - `pnet::packet::Packet` (`ExamplePacket` and `MutableExamplePacket`)
//!      - `pnet::packet::MutablePacket` (`MutableExamplePacket` only)
//!      - `std::fmt::Debug` (`ExamplePacket` and `MutableExamplePacket`)
//!      - `pnet::packet::FromPacket` (`ExamplePacket` and `MutableExamplePacket`)
//!      - `pnet::packet::PacketSize` (`ExamplePacket` and `MutableExamplePacket`)
//!  * An `ExampleIterator` structure, which implements `std::iter::Iterator`, to allow iterating
//!    over vectors of `ExamplePacket` contained within another packet. Used internally.
//!
//! ## Attributes
//!
//! There are a number of attributes which fields may have, these include:
//!
//!  * \#[length_fn = "function_name"]
//!
//!    This attribute is used to enable variable length fields. To specify a variable length field,
//!    it should have the type `Vec<T>`. It must have the `#[length_fn]` (or #[length]) attribute,
//!    which specifies a function name to calculate the length of the field. The signature for the
//!    length function should be
//!    `fn {function_name}<'a>(example_packet: &ExamplePacket<'a>) -> usize`, substituting
//!    `&ExamplePacket<'a>` for the appropriately named packet type for your structure. You may
//!    access whichever fields are required to calculate the length of the field. The returned
//!    value should be a number of bytes that the field uses.
//!
//!    The type contained in the vector may either be one of the primitive types specified in
//!    `pnet_macros::types`, or another structure marked with #[packet], for example
//!    `Vec<Example>`.
//!
//!  * \#[length = "arithmetic expression"]
//!
//!    This attribute is used to enable variable length fields. To specify a variable length field,
//!    it should have the type `Vec<T>`. It must have the `#[length]` (or #[length_fn]) attribute,
//!    which specifies an arithmetic expression to calculate the length of the field. Only field
//!    names, constants, integers, basic arithmetic expressions (+ - * / %) and parentheses are
//!    in the expression. An example would be `#[length = "field_name + CONSTANT - 4]`.
//!
//!    The type contained in the vector may either be one of the primitive types specified in
//!    `pnet_macros::types`, or another structure marked with #[packet], for example
//!    `Vec<Example>`.
//!
//!  * \#[payload]
//!
//!    This attribute specifies the payload associated with the packet. This should specify the
//!    data associated with the packet. It may be used in two places:
//!      - The last field in the packet, in which case it is assumed to use the remaining length of
//!        the buffer containing the packet
//!      - Another location in the packet, in which case the `#[length_fn]` attribute must also be
//!        specified to give the length of the payload.
//!    If the packet has no payload, you must still specify this attribute, but you can provide a
//!    `#[length_fn]` attribute returning zero.
//!
//!  * \#[construct_with(<primitive type>, ...)]
//!
//!    Unfortunately, compiler plugins do not currently have access to type information during the
//!    decoration stage (where all of the above is generated), so this attribute is required. This
//!    must be used for all fields which are neither primitive types, nor vectors of primitive
//!    types. Three things are required when using `#[construct_with]`:
//!      - The field type must have a method `new`, which takes one or more parameters of primitive
//!        types.
//!      - The field must be annotated with the `#[construct_with(...)]` attribute, specifying a
//!        list of types identical to those taken by the `new` method.
//!      - The `pnet::packet::ToPrimitiveValues` trait must be implemented for the field type,
//!        which must return a tuple of the primitive types specified in the parameters to the
//!        `#[construct_with(...)]` attribute, and in the `new` method.

#![deny(missing_docs)]
// #![cfg_attr(feature = "clippy", feature(plugin))]
// #![cfg_attr(feature = "clippy", plugin(clippy))]
// #![cfg_attr(feature = "clippy", allow(let_and_return))]

use proc_macro::TokenStream;
use quote::quote;
use syn::{parse_macro_input, DeriveInput, Visibility};

mod decorator;
mod util;

/// The entry point for the `derive(Packet)` custom derive
#[proc_macro_derive(Packet, attributes(construct_with, length, length_fn, payload))]
pub fn derive_packet(input: TokenStream) -> TokenStream {
    let ast = parse_macro_input!(input as DeriveInput);
    // ensure struct is public
    match ast.vis {
        Visibility::Public(_) => (),
        _ => {
            let ts = syn::Error::new(ast.ident.span(), "#[packet] structs must be public")
                .to_compile_error();
            return ts.into();
        }
    }
    let name = &ast.ident;
    let s = match &ast.data {
        syn::Data::Struct(ref s) => decorator::generate_packet(s, name.to_string()),
        _ => panic!("Only structs are supported"),
    };
    s.into()
}

/// The entry point for the `packet` proc_macro_attribute
#[proc_macro_attribute]
pub fn packet(_attrs: TokenStream, code: TokenStream) -> TokenStream {
    // let _attrs = parse_macro_input!(attrs as AttributeArgs);
    let input = parse_macro_input!(code as DeriveInput);
    // enhancement: if input already has Clone and/or Debug, do not add them
    let s = quote! {
        #[derive(::pnet_macros::Packet, Clone, Debug)]
        #[allow(unused_attributes)]
        #input
    };
    s.into()
}
