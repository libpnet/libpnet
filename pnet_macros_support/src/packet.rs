// Copyright (c) 2014, 2015, 2017 Robert Clipsham <robert@octarineparrot.com>
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

//! Packet helpers for `pnet_macros`.

use pnet_base;
use std::ops::{Deref, DerefMut, Index, IndexMut, Range, RangeFrom, RangeFull, RangeTo};

/// Represents a generic network packet.
pub trait Packet {
    /// Retrieve the underlying buffer for the packet.
    fn packet(&self) -> &[u8];

    /// Retrieve the payload for the packet.
    fn payload(&self) -> &[u8];
}

/// Represents a generic, mutable, network packet.
pub trait MutablePacket: Packet {
    /// Retreive the underlying, mutable, buffer for the packet.
    fn packet_mut(&mut self) -> &mut [u8];

    /// Retreive the mutable payload for the packet.
    fn payload_mut(&mut self) -> &mut [u8];

    /// Initialize this packet by cloning another.
    fn clone_from<T: Packet>(&mut self, other: &T) {
        use std::ptr;

        assert!(self.packet().len() >= other.packet().len());
        unsafe {
            ptr::copy_nonoverlapping(
                other.packet().as_ptr(),
                self.packet_mut().as_mut_ptr(),
                other.packet().len(),
            );
        }
    }
}

/// Used to convert on-the-wire packets to their #[packet] equivalent.
pub trait FromPacket: Packet {
    /// The type of the packet to convert from.
    type T;

    /// Converts a wire-format packet to #[packet] struct format.
    fn from_packet(&self) -> Self::T;
}

/// Used to find the calculated size of the packet. This is used for occasions where the underlying
/// buffer is not the same length as the packet itself.
pub trait PacketSize: Packet {
    /// Get the calculated size of the packet.
    fn packet_size(&self) -> usize;
}

macro_rules! impl_index {
    ($t:ident, $index_t:ty, $output_t:ty) => {
        impl<'p> Index<$index_t> for $t<'p> {
            type Output = $output_t;

            fn index(&self, index: $index_t) -> &$output_t {
                &self.as_slice().index(index)
            }
        }
    };
}

macro_rules! impl_index_mut {
    ($t:ident, $index_t:ty, $output_t:ty) => {
        impl<'p> IndexMut<$index_t> for $t<'p> {
            fn index_mut(&mut self, index: $index_t) -> &mut $output_t {
                self.as_mut_slice().index_mut(index)
            }
        }
    };
}

/// Packet data.
#[derive(PartialEq)]
pub enum PacketData<'p> {
    /// A packet owns its contents.
    Owned(Vec<u8>),
    /// A packet borrows its contents.
    Borrowed(&'p [u8]),
}

impl<'p> PacketData<'p> {
    /// Get a slice of the packet data.
    pub fn as_slice(&self) -> &[u8] {
        match self {
            &PacketData::Owned(ref data) => data.deref(),
            &PacketData::Borrowed(ref data) => data,
        }
    }

    /// No-op - returns `self`.
    pub fn to_immutable(self) -> PacketData<'p> {
        self
    }

    /// A length of the packet data.
    pub fn len(&self) -> usize {
        self.as_slice().len()
    }
}

impl_index!(PacketData, usize, u8);
impl_index!(PacketData, Range<usize>, [u8]);
impl_index!(PacketData, RangeTo<usize>, [u8]);
impl_index!(PacketData, RangeFrom<usize>, [u8]);
impl_index!(PacketData, RangeFull, [u8]);

/// Mutable packet data.
#[derive(PartialEq)]
pub enum MutPacketData<'p> {
    /// Owned mutable packet data.
    Owned(Vec<u8>),
    /// Borrowed mutable packet data.
    Borrowed(&'p mut [u8]),
}

impl<'p> MutPacketData<'p> {
    /// Get packet data as a slice.
    pub fn as_slice(&self) -> &[u8] {
        match self {
            &MutPacketData::Owned(ref data) => data.deref(),
            &MutPacketData::Borrowed(ref data) => data,
        }
    }

    /// Get packet data as a mutable slice.
    pub fn as_mut_slice(&mut self) -> &mut [u8] {
        match self {
            &mut MutPacketData::Owned(ref mut data) => data.deref_mut(),
            &mut MutPacketData::Borrowed(ref mut data) => data,
        }
    }

    /// Get an immutable version of packet data.
    pub fn to_immutable(self) -> PacketData<'p> {
        match self {
            MutPacketData::Owned(data) => PacketData::Owned(data),
            MutPacketData::Borrowed(data) => PacketData::Borrowed(data),
        }
    }

    /// Get a length of data in the packet.
    pub fn len(&self) -> usize {
        self.as_slice().len()
    }
}

impl_index!(MutPacketData, usize, u8);
impl_index!(MutPacketData, Range<usize>, [u8]);
impl_index!(MutPacketData, RangeTo<usize>, [u8]);
impl_index!(MutPacketData, RangeFrom<usize>, [u8]);
impl_index!(MutPacketData, RangeFull, [u8]);

impl_index_mut!(MutPacketData, usize, u8);
impl_index_mut!(MutPacketData, Range<usize>, [u8]);
impl_index_mut!(MutPacketData, RangeTo<usize>, [u8]);
impl_index_mut!(MutPacketData, RangeFrom<usize>, [u8]);
impl_index_mut!(MutPacketData, RangeFull, [u8]);

/// Used to convert a type to primitive values representing it.
pub trait PrimitiveValues {
    /// A tuple of types, to represent the current value.
    type T;

    /// Convert a value to primitive types representing it.
    fn to_primitive_values(&self) -> Self::T;
}

impl PrimitiveValues for pnet_base::MacAddr {
    type T = (u8, u8, u8, u8, u8, u8);
    fn to_primitive_values(&self) -> (u8, u8, u8, u8, u8, u8) {
        (self.0, self.1, self.2, self.3, self.4, self.5)
    }
}

impl PrimitiveValues for ::std::net::Ipv4Addr {
    type T = (u8, u8, u8, u8);
    fn to_primitive_values(&self) -> (u8, u8, u8, u8) {
        let octets = self.octets();

        (octets[0], octets[1], octets[2], octets[3])
    }
}

impl PrimitiveValues for ::std::net::Ipv6Addr {
    type T = (u16, u16, u16, u16, u16, u16, u16, u16);
    fn to_primitive_values(&self) -> (u16, u16, u16, u16, u16, u16, u16, u16) {
        let segments = self.segments();

        (
            segments[0],
            segments[1],
            segments[2],
            segments[3],
            segments[4],
            segments[5],
            segments[6],
            segments[7],
        )
    }
}
