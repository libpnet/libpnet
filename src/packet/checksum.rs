//! rfc1071 checksum abstraction

/// Calculates rfc1071 checksum value
pub fn rfc1071_checksum(packet: &[u8], initial: u32) -> u16 {
    let length = packet.len() - 1;
    let mut sum = initial;

    let mut i = 0;
    while i < length {
        let word = (packet[i] as u32) << 8 | packet[i + 1] as u32;
        sum = sum + word;
        i = i + 2;
    }
    if packet.len()%2 == 1 {
        sum = sum + (packet[length] as u32) << 8
    }
    while sum >> 16 != 0 {
        sum = (sum >> 16) + (sum & 0xFFFF);
    }
    return !sum as u16;
}
