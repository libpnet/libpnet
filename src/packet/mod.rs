// Copyright (c) 2014, 2015 Robert Clipsham <robert@octarineparrot.com>
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

//! Provides interfaces for interacting with packets and headers

#![macro_use]

pub enum PseudoHeaderIPv6IPv4 {
    Ipv6PseudoHeader,
    Ipv4PseudoHeader,
}

/// Used to help compute checksum value of packet
pub trait PseudoHeader {
    /// Converts IPv4 or IPv6 specific fields to a checksum value
    // Should we add a "next_header" u16 argument for ipv6 packets with extension header?
    //
    // According to rfc 2460, regarding the "next header" field in the ipv6 pseudo header:
    //
    // The Next Header value in the pseudo-header identifies the
    // upper-layer protocol (e.g., 6 for TCP, or 17 for UDP).  It will
    // differ from the Next Header value in the IPv6 header if there
    // are extension headers between the IPv6 header and the upper-
    // layer header.
    //
    fn populate(&self, buffer: &mut [u8], inner_packet_length: Option<usize>) -> PseudoHeaderIPv6IPv4 ;
}

/// Represents a generic network packet
pub trait Packet {
    /// Retreive the underlying buffer for the packet
    fn packet(&self) -> &[u8];

    /// Retreive the payload for the packet
    fn payload(&self) -> &[u8];
}

/// Represents a generic, mutable, network packet
pub trait MutablePacket : Packet {
    /// Retreive the underlying, mutable, buffer for the packet
    fn packet_mut(&mut self) -> &mut [u8];

    /// Retreive the mutable payload for the packet
    fn payload_mut(&mut self) -> &mut [u8];

    /// Initialize this packet by cloning another
    fn clone_from<T: Packet>(&mut self, other: &T) {
        use std::ptr;

        assert!(self.packet().len() >= other.packet().len());
        unsafe {
            ptr::copy_nonoverlapping(other.packet().as_ptr(),
                                     self.packet_mut().as_mut_ptr(),
                                     other.packet().len());
        }
    }
}

/// Used to convert on-the-wire packets to their #[packet] equivalent
pub trait FromPacket : Packet {
    /// The type of the packet to convert from
    type T;

    /// Converts a wire-format packet to #[packet] struct format
    fn from_packet(&self) -> Self::T;
}

/// Used to find the calculated size of the packet. This is used for occasions where the underlying
/// buffer is not the same length as the packet itself.
pub trait PacketSize : Packet {
    /// Get the calculated size of the packet
    fn packet_size(&self) -> usize;
}

/// Used to convert a type to primitive values representing it
pub trait PrimitiveValues {
    /// A tuple of types, to represent the current value
    type T;

    /// Convert a value to primitive types representing it
    fn to_primitive_values(&self) -> Self::T;
}

impl PrimitiveValues for ::std::net::Ipv4Addr {
    type T = (u8, u8, u8, u8);
    fn to_primitive_values(&self) -> (u8, u8, u8, u8) {
        let octets = self.octets();

        (octets[0], octets[1], octets[2], octets[3])
    }
}

impl PrimitiveValues for ::std::net::Ipv6Addr {
    type T = (u16, u16, u16, u16, u16, u16, u16, u16);
    fn to_primitive_values(&self) -> (u16, u16, u16, u16, u16, u16, u16, u16) {
        let segments = self.segments();

        (segments[0],
         segments[1],
         segments[2],
         segments[3],
         segments[4],
         segments[5],
         segments[6],
         segments[7])
    }
}

pub mod ethernet;
pub mod ip;
pub mod ipv4;
pub mod ipv6;
pub mod udp;
