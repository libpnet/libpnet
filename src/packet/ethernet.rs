// Copyright (c) 2014 Robert Clipsham <robert@octarineparrot.com>
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

//! Ethernet packet abstraction

use std::fmt;
use packet::{Packet, MutablePacket};
use util::MacAddr;

/// A structure which represents an Ethernet header
pub struct EthernetHeader<'p> {
    packet: &'p [u8],
}
impl<'p> Copy for EthernetHeader<'p> {}

// FIXME This should probably be a macro
// FIXME What should be the proper behaviour of this?
//        - Compare whole packet?
//        - Compare just header?
//        - Compare non-volatile bits of header? (ignore ttl etc)
impl<'p> PartialEq for EthernetHeader<'p> {
    fn eq(&self, other: &EthernetHeader) -> bool {
        if self.packet.len() != other.packet.len() {
            return false;
        }
        for (b1, b2) in self.packet.iter()
                        .zip(other.packet.iter()) {
            if b1 != b2 {
                return false;
            }
        }
        return true;
    }
}
impl<'p> Eq for EthernetHeader<'p> {}

impl<'p> fmt::String for EthernetHeader<'p> {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt,
               "EthernetHeader {{ destination: {}, source: {}, ethertype: {} }}",
               self.get_destination(),
               self.get_source(),
               self.get_ethertype())
    }
}

// NOTE Copy/pasted from above.
impl<'p> fmt::String for MutableEthernetHeader<'p> {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(fmt,
               "MutableEthernetHeader {{ destination: {}, source: {}, ethertype: {} }}",
               self.get_destination(),
               self.get_source(),
               self.get_ethertype())
    }
}

/// A structure representing an Ethernet header which can be mutated
pub struct MutableEthernetHeader<'p> {
    packet: &'p mut [u8],
}

impl<'a> Packet for EthernetHeader<'a> {
    #[inline(always)]
    fn packet<'p>(&'p self) -> &'p [u8] { self.packet }

    #[inline(always)]
    fn payload<'p>(&'p self) -> &'p [u8] { self.packet.slice_from(14) }
}

impl<'a> Packet for MutableEthernetHeader<'a> {
    #[inline(always)]
    fn packet<'p>(&'p self) -> &'p [u8] { self.packet.as_slice() }

    #[inline(always)]
    fn payload<'p>(&'p self) -> &'p [u8] { self.packet.slice_from(14) }
}

impl<'a> MutablePacket for MutableEthernetHeader<'a> {
    #[inline(always)]
    fn packet_mut<'p>(&'p mut self) -> &'p mut [u8] { self.packet.as_mut_slice() }

    #[inline(always)]
    fn payload_mut<'p>(&'p mut self) -> &'p mut [u8] { self.packet.slice_from_mut(14) }
}

/// A trait implemented by anything which provides the ability to retrieve
/// fields of an Ethernet packet
pub trait EthernetPacket : Packet {
    /// Get the destination address for an Ethernet packet
    fn get_destination(&self) -> MacAddr {
        MacAddr(
            self.packet()[0],
            self.packet()[1],
            self.packet()[2],
            self.packet()[3],
            self.packet()[4],
            self.packet()[5]
        )
    }

    /// Get the source address for an Ethernet packet
    fn get_source(&self) -> MacAddr {
        MacAddr(
            self.packet()[6],
            self.packet()[7],
            self.packet()[8],
            self.packet()[9],
            self.packet()[10],
            self.packet()[11]
        )
    }

    /// Get the Ethertype field of an Ethernet packet
    fn get_ethertype(&self) -> EtherType {
        EtherType(((self.packet()[12] as u16) << 8) | (self.packet()[13] as u16))
    }
}

impl<'p> EthernetPacket for EthernetHeader<'p> {}
impl<'p> EthernetPacket for MutableEthernetHeader<'p> {}

impl<'p> EthernetHeader<'p> {
    /// Construct a new Ethernet header backed by the given buffer with the
    /// provided offset
    pub fn new(packet: &'p [u8]) -> EthernetHeader<'p> {
        EthernetHeader { packet: packet }
    }
}

impl<'p> MutableEthernetHeader<'p> {
    /// Construct a new mutable Ethernet header backed by the given buffer with
    /// the provided offset
    pub fn new(packet: &'p mut [u8]) -> MutableEthernetHeader<'p> {
        MutableEthernetHeader { packet: packet }
    }

    /// Set the source address for an Ethernet packet
    pub fn set_source(&mut self, mac: MacAddr) {
        match mac {
            MacAddr(a, b, c, d, e, f) => {
                self.packet[6] = a;
                self.packet[7] = b;
                self.packet[8] = c;
                self.packet[9] = d;
                self.packet[10] = e;
                self.packet[11] = f;
            }
        }
    }

    /// Set the destination address for an Ethernet packet
    pub fn set_destination(&mut self, mac: MacAddr) {
        match mac {
            MacAddr(a, b, c, d, e, f) => {
                self.packet[0] = a;
                self.packet[1] = b;
                self.packet[2] = c;
                self.packet[3] = d;
                self.packet[4] = e;
                self.packet[5] = f;
            }
        }
    }

    /// Set the Ethertype for an Ethernet packet
    pub fn set_ethertype(&mut self, EtherType(ethertype): EtherType) {
        self.packet[12] = (ethertype >> 8) as u8;
        self.packet[13] = (ethertype & 0xFF) as u8;
    }
}

#[test]
fn ethernet_header_test() {
    let mut packet = [0u8; 14];
    {
        let mut ethernet_header = MutableEthernetHeader::new(packet.as_mut_slice());

        let source = MacAddr(0x12, 0x34, 0x56, 0x78, 0x9a, 0xbc);
        ethernet_header.set_source(source);
        assert_eq!(ethernet_header.get_source(), source);

        let dest = MacAddr(0xde, 0xf0, 0x12, 0x34, 0x45, 0x67);
        ethernet_header.set_destination(dest);
        assert_eq!(ethernet_header.get_destination(), dest);

        ethernet_header.set_ethertype(EtherTypes::Ipv6);
        assert_eq!(ethernet_header.get_ethertype(), EtherTypes::Ipv6);
    }

    let ref_packet = [0xde, 0xf0, 0x12, 0x34, 0x45, 0x67, /* destination */
                      0x12, 0x34, 0x56, 0x78, 0x9a, 0xbc, /* source */
                      0x86, 0xdd /* ethertype */];
    assert_eq!(ref_packet.as_slice(), packet.as_slice());
}

/// EtherTypes defined at:
/// http://www.iana.org/assignments/ieee-802-numbers/ieee-802-numbers.xhtml
/// These values should be used in the Ethernet EtherType field
///
/// FIXME Should include all
/// A handful of these have been selected since most are archaic and unused.
#[allow(non_snake_case)]
#[allow(non_upper_case_globals)]
pub mod EtherTypes {
    use packet::ethernet::EtherType;

    pub const Ipv4: EtherType      = EtherType(0x0800);
    pub const Arp: EtherType       = EtherType(0x0806);
    pub const WakeOnLan: EtherType = EtherType(0x0842);
    pub const Rarp: EtherType      = EtherType(0x8035);
    pub const Ipv6: EtherType      = EtherType(0x86DD);
}

/// Represents the Ethernet ethertype field.
#[derive(Show, PartialEq, Eq, PartialOrd, Ord)]
pub struct EtherType(pub u16);
impl Copy for EtherType {}

impl fmt::String for EtherType {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        let EtherType(ethertype) = *self;
        write!(fmt, "{}", ethertype)
    }
}

