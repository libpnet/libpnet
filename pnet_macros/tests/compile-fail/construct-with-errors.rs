// Copyright (c) 2021 Pierre Chifflier <chifflier@wzdftpd.net>
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

extern crate pnet_macros;
extern crate pnet_macros_support;
use pnet_macros::packet;

#[derive(Clone, Debug)]
pub struct Toto{
    dummy: u16,
}

#[packet]
pub struct PacketU16 {
    #[construct_with()] //~ ERROR #[construct_with] must have at least one argument
    banana: Toto,
    #[payload]
    payload: Vec<u8>,
}

#[packet]
pub struct PacketU16B {
    #[construct_with("test")] //~ ERROR #[construct_with] should be of the form #[construct_with(<types>)]
    banana: Toto,
    #[payload]
    payload: Vec<u8>,
}

fn main() {}
