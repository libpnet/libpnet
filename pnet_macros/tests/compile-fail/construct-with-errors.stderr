error: #[construct_with] must have at least one argument
  --> $DIR/construct-with-errors.rs:20:7
   |
20 |     #[construct_with()] //~ ERROR #[construct_with] must have at least one argument
   |       ^^^^^^^^^^^^^^

error: #[construct_with] should be of the form #[construct_with(<types>)]
  --> $DIR/construct-with-errors.rs:28:22
   |
28 |     #[construct_with("test")] //~ ERROR #[construct_with] should be of the form #[construct_with(<types>)]
   |                      ^^^^^^
